autoscale/src/main/java/com/sequenceiq/periscope/service/configuration/CloudInstanceTypeService.java

```diff
@@ -0,0 +1,84 @@
+package com.sequenceiq.periscope.service.configuration;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.sequenceiq.cloudbreak.client.CloudbreakInternalCrnClient;
+import com.sequenceiq.cloudbreak.common.json.JsonUtil;
+import com.sequenceiq.cloudbreak.common.mappable.CloudPlatform;
+import com.sequenceiq.cloudbreak.service.CloudbreakResourceReaderService;
+import com.sequenceiq.periscope.model.CloudInstanceType;
+import com.sequenceiq.periscope.utils.FileReaderUtils;
+
+@Component
+public class CloudInstanceTypeService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileReaderUtils.class);
+
+    @Inject
+    private CloudbreakResourceReaderService cloudbreakResourceReaderService;
+
+    @Inject
+    private CloudbreakInternalCrnClient internalCrnClient;
+
+    private Map<String, CloudInstanceType> awsInstanceTypes = new HashMap<>();
+
+    private Map<String, CloudInstanceType> azureInstanceTypes = new HashMap<>();
+
+    @PostConstruct
+    public void init() {
+        readCloudInstanceTypes("aws", awsInstanceTypes);
+        readCloudInstanceTypes("azure", azureInstanceTypes);
+    }
+
+    public Optional<CloudInstanceType> getCloudVMInstanceType(CloudPlatform cloudPlatform, String hostGroupInstanceType) {
+        CloudInstanceType cloudVmType;
+        switch (cloudPlatform) {
+            case AWS:
+                cloudVmType = awsInstanceTypes.get(hostGroupInstanceType);
+                break;
+            case AZURE:
+                cloudVmType = azureInstanceTypes.get(hostGroupInstanceType);
+                break;
+            default:
+                cloudVmType = null;
+        }
+        return Optional.ofNullable(cloudVmType);
+    }
+
+    private void readCloudInstanceTypes(String cloudPlatform, Map<String, CloudInstanceType> vmInstanceTypes) {
+        try {
+            String cloudInstanceTypeConfig = cloudbreakResourceReaderService.resourceDefinition(cloudPlatform, "vm");
+
+            TypeReference<HashMap<String, Set<CloudInstanceType>>> typeRef = new TypeReference<>() { };
+            Set<CloudInstanceType> cloudInstanceTypes = JsonUtil.readValue(cloudInstanceTypeConfig, typeRef).get("items");
+
+            for (CloudInstanceType cloudInstanceType : cloudInstanceTypes) {
+                vmInstanceTypes.put(cloudInstanceType.getInstanceName(), cloudInstanceType);
+            }
+
+            Set<String> configuredCloudInstances = cloudInstanceTypes.stream().map(CloudInstanceType::getInstanceName).collect(Collectors.toSet());
+            Set<String> cbSupportedInstances = internalCrnClient.withInternalCrn()
```

Comment:
Can the CB returned instance types be used directly? Why define them again in Periscope.
Alternately, if a library exists which can do this conversion - Periscope can rely on the DB entry directly.

Reply:
commented above. I dont think this is required in DB, it is stored in resourcedefinitions, so that it can be easily updated without generating sql.

Reply:
Sorry. I meant the Cloudbreak DB. So specifically.
- An instance type is retrieved from the hostgroup info in the Cloudbreak DB.
- Periscope uses a library (which at the moment is backed by aws-, azure-*.json in the repo).
The intent being never to talk to Cloudbreak.

That is pretty much the case already. This communication to Cloudbreak to fetch instance types is only for error checking.
On the error checking - I think it makes sense to get Periscope to crash if information is not available for an instance type. That forces errors to be caught early.

If this will be maintained as these json files with look up - I think we should actually add an Integration test which ensures that what CB supports and Periscope knows about match, to fail even earlier. (Integration test is a follow up jira).

Reply:
agree. i didnt crash periscope on error checking since in that case, periscope will never start if cloudbreak is not running. agree enabling integration testing is the correct approach here. will create a jira for that.

Reply:
Makes sense.